feat: Enhance AutoTuner inference path and code readability

hyukn opened the pull request:
Fix AutoTuner warmup request generating.
* The current warmup phase creates one request, which is insufficient for the warmup to cover the max_num_tokens. Revise the warmup phase to a batch of requests to cover the max_num_tokens to eliminate potential fallback cases.

Refactor AutoTuner API and reduce host overhead.
* Refine (min, opt, max) values of optimization profile setup for get_valid_tactics to achieve the correct canImplement definition.
* Refine cache key assembly process to reduce host overhead and simplify API.
* Fix lru_cache usage to reduce host overhead.
* Move tuning config initialization as one-time object in tunable runner to reduce host overhead.

Improve tuning config readability. Use dataclass to define tuning config.

hyukn commented:
/bot run

tensorrt-cicd commented:
[PR_Github #5806](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/5806/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #5806](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/5806/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #4253](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/4253/) completed with status: 'SUCCESS'

hyukn commented:
/bot run

hyukn commented:
/bot run

tensorrt-cicd commented:
[PR_Github #5928](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/5928/) [ run ] triggered by Bot

hyukn commented:
/bot run

tensorrt-cicd commented:
[PR_Github #5958](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/5958/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #5958](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/5958/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #4358](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/4358/) completed with status: 'FAILURE'

hyukn commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #6105](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/6105/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #6126](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/6126/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #6126](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/6126/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #4475](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/4475/) completed with status: 'FAILURE'

hyukn commented:
/bot run

tensorrt-cicd commented:
[PR_Github #6223](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/6223/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #6223](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/6223/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #4547](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/4547/) completed with status: 'FAILURE'

hyukn commented:
/bot run --add-multi-gpu-test

tensorrt-cicd commented:
[PR_Github #6242](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/6242/) [ run ] triggered by Bot

hyukn commented:
/bot run --add-multi-gpu-test

tensorrt-cicd commented:
[PR_Github #6270](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/6270/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #6242](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/6242/) [ run ] completed with state `ABORTED`

tensorrt-cicd commented:
[PR_Github #6270](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/6270/) [ run ] completed with state `FAILURE`
[/LLM/main/L0_MergeRequest_PR pipeline #4582](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/4582/) completed with status: 'FAILURE'

hyukn commented:
/bot run --add-multi-gpu-test

tensorrt-cicd commented:
[PR_Github #6286](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/6286/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #6286](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/6286/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #4594](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/4594/) completed with status: 'FAILURE'

hlu1 commented on tensorrt_llm/_torch/autotuner.py:
```diff
@@ -13,31 +13,46 @@
+
+
+@dataclass
+class Constraint:
+    tensor_idx: int
+    dim_idx: int
+    shape_infer_func: Callable
```
```suggestion
class Constraint:
    input_idx: int
    dim_idx: int
    shape_infer_func: Callable
```

hlu1 commented on tensorrt_llm/_torch/autotuner.py:
```diff
@@ -13,31 +13,46 @@
 
 
+@dataclass(frozen=True)
+class DynamicTensor:
+    tensor_idx: int
+    dim_idx: int
+    generating_rule: Union[Tuple[int], Callable]
+    mapping_rule: Callable
```
```suggestion
class DynamicTensorSpec:
    input_idx: int
    dim_idx: int
    gen_tuning_buckets: Union[Tuple[int], Callable]
    map_to_tuning_buckets: Callable
```
Some suggestions for better naming. This class not a Tensor. It's more of a tuning spec.

hyukn replied:
Thanks for the suggestions. Will apply it.

hyukn commented on tensorrt_llm/_torch/pyexecutor/model_engine.py:
```diff
@@ -531,6 +531,48 @@ def get_torch_compile_warmup_request(batch_size,
+
+            # We create multiple requests with the possible maximum sequence length
+            # which should be num_tokens_per_request computed with the constraints.
+            request_num_pre = self.max_num_tokens // num_tokens_per_request
```
This is to indicate the request number precedent to the real request number because remainder number will determine if an extra request is required. Will make these variables more comprehensible.

tensorrt-cicd commented:
[PR_Github #6358](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/6358/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #6358](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/6358/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #4643](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/4643/) completed with status: 'FAILURE'

hyukn committed:
Autotuner API refactor and optimization.
Fix AutoTuner warmup request generating.
* The current warmup phase creates one request, which is insufficient for the warmup to cover the max_num_tokens. Revise the warmup phase to a batch of requests to cover the max_num_tokens to eliminate potential fallback cases.

Refactor AutoTuner API and reduce host overhead.
* Refine (min, opt, max) values of optimization profile setup for get_valid_tactics to achieve correct canImplement definition.
* Refine cache key assembly process to reduce host overhead and simplify API.
* Fix lru_cache usage to reduce host overhead.
* Move tuning config initialization as onetime object in tunable runner to reduce host overhead.

Improve tuning config readability. Use dataclass to define tuning config.

Signed-off-by: Yukun He <23156053+hyukn@users.noreply.github.com>

hyukn commented:
/bot run --add-multi-gpu-test

hyukn commented:
/bot run --add-multi-gpu-test

hyukn commented:
/bot run --add-multi-gpu-test

tensorrt-cicd commented:
[PR_Github #6973](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/6973/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #6973](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/6973/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #5060](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/5060/) completed with status: 'FAILURE'

hyukn commented:
/bot run --add-multi-gpu-test

tensorrt-cicd commented:
[PR_Github #7117](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/7117/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #7117](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/7117/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #5142](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/5142/) completed with status: 'SUCCESS'

hyukn commented on tensorrt_llm/_torch/autotuner.py:
```diff
@@ -107,7 +138,8 @@ class FakeTensor:
     @abstractmethod
-    def get_valid_tactics(self, inputs: List[FakeTensor]) -> List[int]:
+    def get_valid_tactics(self, inputs: List[torch.Tensor],
+                          profile: OptimizationProfile) -> List[int]:
```
Then we need to reassemble fake tensors and pass them to the method. I am going to see a proper place to trigger this happening.

hyukn commented:
/bot run

tensorrt-cicd commented:
[PR_Github #7304](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/7304/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #7304](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/7304/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #5292](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/5292/) completed with status: 'SUCCESS'
Pipeline passed with automatic retried tests. Check the [rerun report](https://urm.nvidia.com/artifactory/sw-tensorrt-generic/llm-artifacts//LLM/main/L0_MergeRequest_PR/5292/test-results/rerun_report.html) for details.
